lib/ecto/changeset.ex

```diff
@@ -1523,10 +1523,10 @@ defmodule Ecto.Changeset do
 
   defp cast_merge(cs1, cs2) do
     new_params = (cs1.params || cs2.params) && Map.merge(cs1.params || %{}, cs2.params || %{})
+    new_types = Map.merge(cs1.types, cs2.types)
```

Comment:
It seems the assumption is that some people were initializing the changeset directly without the proper API. I don't think we need to support that.